File: python/res/enkf/rng_config.py
#  Copyright (C) 2017  Equinor ASA, Norway.
#
#  The file 'rng_config.py' is part of ERT - Ensemble based Reservoir Tool.
#
#  ERT is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ERT is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE.
#
#  See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
#  for more details.


from cwrap import BaseCClass

from res import ResPrototype
from res.enkf import ConfigKeys


class RNGConfig(BaseCClass):
    TYPE_NAME = "rng_config"
    _alloc = ResPrototype("void* rng_config_alloc(config_content)", bind=False)
    _alloc_full = ResPrototype("void* rng_config_alloc_full(char*)", bind=False)
    _free = ResPrototype("void rng_config_free(rng_config)")
    _rng_alg_type = ResPrototype("rng_alg_type_enum rng_config_get_type(rng_config)")
    _random_seed = ResPrototype("char* rng_config_get_random_seed(rng_config)")

    def __init__(self, config_content=None, config_dict=None):
        if config_content and config_dict:
            raise ValueError("RNGConfig can not be instantiated with both config types")

        if not (config_content or config_dict):
            raise ValueError(
                "RNGConfig can not be instantiated without any config objects"
            )

        if config_content:
            c_ptr = self._alloc(config_content)
        elif config_dict:
            random_seed = config_dict.get(ConfigKeys.RANDOM_SEED)
            c_ptr = self._alloc_full(random_seed)
        else:
            c_ptr = None

        if c_ptr is None:
            raise ValueError("Failed to construct RNGConfig instance")

        super(RNGConfig, self).__init__(c_ptr)

    @property
    def alg_type(self):
        return self._rng_alg_type()

    @property
    def random_seed(self):
        return self._random_seed()

    def free(self):
        self._free()

    def __eq__(self, other):
        if self.random_seed != other.random_seed:
            return False

        if self.alg_type != other.alg_type:
            return False

        return True

File: lib/enkf/rng_config.cpp
/*
   Copyright (C) 2011  Equinor ASA, Norway.

   The file 'rng_config.c' is part of ERT - Ensemble based Reservoir Tool.

   ERT is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.

   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
   for more details.
*/

#include <stdlib.h>

#include <ert/util/mzran.h>
#include <ert/util/util.h>
#include <ert/util/test_util.h>

#include <ert/config/config_parser.hpp>
#include <ert/config/config_schema_item.hpp>

#include <ert/res_util/res_log.hpp>

#include <ert/enkf/rng_config.hpp>
#include <ert/enkf/rng_manager.hpp>
#include <ert/enkf/config_keys.hpp>
#include <ert/enkf/enkf_defaults.hpp>
#include <ert/enkf/model_config.hpp>


struct rng_config_struct {
  rng_alg_type      type;
  char            * random_seed;
};


static void rng_config_set_random_seed(rng_config_type * rng_config, const char * random_seed) {
  rng_config->random_seed = util_realloc_string_copy(rng_config->random_seed, random_seed);
}

static char * rng_config_alloc_formatted_random_seed(const rng_config_type * rng_config) {
  unsigned int * fseed = (unsigned int*) malloc(RNG_STATE_SIZE * sizeof(unsigned int*));

  int seed_len = strlen(rng_config->random_seed);
  int seed_pos = 0;
  for (int i = 0; i < RNG_STATE_SIZE; ++i) {
    fseed[i] = 0;
    for (int k = 0; k < RNG_STATE_DIGITS; ++k) {
      fseed[i] *= 10;
      fseed[i] += rng_config->random_seed[seed_pos] - '0';
      seed_pos = (seed_pos+1) % seed_len;
    }
  }

  return (char *) fseed;
}

const char * rng_config_get_random_seed(const rng_config_type * rng_config) {
  return rng_config->random_seed;
}

void rng_config_set_type( rng_config_type * rng_config , rng_alg_type type) {
  rng_config->type = type;
}

rng_alg_type rng_config_get_type(const rng_config_type * rng_config ) {
  return rng_config->type;
}

static rng_config_type * rng_config_alloc_default(void) {
  rng_config_type * rng_config = (rng_config_type *)util_malloc( sizeof * rng_config);

  rng_config_set_type( rng_config , MZRAN );  /* Only type ... */
  rng_config->random_seed = NULL;

  return rng_config;
}

rng_config_type * rng_config_alloc_load_user_config(const char * user_config_file) {
  config_parser_type * config_parser = config_alloc();
  config_content_type * config_content = NULL;
  if(user_config_file)
    config_content = model_config_alloc_content(user_config_file, config_parser);

  rng_config_type * rng_config = rng_config_alloc(config_content);

  config_content_free(config_content);
  config_free(config_parser);

  return rng_config;
}

rng_config_type * rng_config_alloc(const config_content_type * config_content) {
  rng_config_type * rng_config = rng_config_alloc_default();

  if(config_content)
    rng_config_init(rng_config, config_content);

  return rng_config;
}
rng_config_type * rng_config_alloc_full(const char * random_seed) {
  rng_config_type * rng_config = rng_config_alloc_default();
  rng_config->random_seed = util_realloc_string_copy(rng_config->random_seed, random_seed);

  return rng_config;
}


void rng_config_free( rng_config_type * rng) {
  free( rng->random_seed );
  free( rng );
}


rng_manager_type * rng_config_alloc_rng_manager( const rng_config_type * rng_config ) {
  rng_manager_type * rng_manager;

  if (rng_config->random_seed) {
    char * formatted_seed = rng_config_alloc_formatted_random_seed(rng_config);
    rng_manager = rng_manager_alloc(formatted_seed);
  } else {
    rng_manager = rng_manager_alloc_random( );
  }

  rng_manager_log_state(rng_manager);

  return rng_manager;
}


/*****************************************************************/

void rng_config_add_config_items( config_parser_type * parser ) {
  config_add_key_value(parser, RANDOM_SEED_KEY, false, CONFIG_STRING);
}


void rng_config_init(rng_config_type * rng_config, const config_content_type * config_content) {
  if(config_content_has_item(config_content, RANDOM_SEED_KEY)) {
    const char * random_seed = config_content_get_value(config_content, RANDOM_SEED_KEY);
    rng_config_set_random_seed(rng_config, random_seed);
    res_log_fcritical("Using RANDOM_SEED: %s", random_seed);
  }
}
